### PR TITLE
New tab overflow

### PR DESCRIPTION
**Pull request by dsmmcken**
This change makes tabs overflow entirely different. Tabs now occupy one overflow: hidden scroll box, that gains scroll on wheel/arrow buttons/dragging tab. Dropdown is now clone of your current tab list when opened, that is filterable. This makes managing a larger number of tabs in a stack easier, and is more inline with how apps like firefox/vs code present tab overflow behaviour.

Significant changes that need review attention:
header.js
stack.js
tab.js

### CODE REVIEW

File: src/js/controls/Tab.js [hunk 1]
```diff
@@ -193,17 +198,26 @@ lm.utils.copy( lm.controls.Tab.prototype, {
 	 */
 	_onTabClick: function( event ) {
 		// left mouse button or tap
-    	if( event.button === 0 ) {
+    	if( !event || event.button === 0 ) {
```
Comment (illumon-public) [inline src/js/controls/Tab.js:201]: When could this event be null?

File: src/js/controls/Tab.js [hunk 2]
```diff
@@ -193,17 +198,26 @@ lm.utils.copy( lm.controls.Tab.prototype, {
...
-				
 			}
+
+			// might have been called from the dropdown
+			this.header._hideAdditionalTabsDropdown();
+
+			// makes sure clicked tabs scrollintoview (either those partially offscreen or in dropdown)
+			this.element.get(0).scrollIntoView({
+				inline: 'nearest',
+				// behavior: 'smooth'
```
Comment (illumon-public) [inline src/js/controls/Tab.js]: You can just delete this line

File: src/js/controls/Tab.js (hunk 1 above)
Comment (dsmmcken) [reply src/js/controls/Tab.js:201]: In header.js on key.Enter I re-used the _onTabClick function to set the active tab... a little messy.

`this.tabs[this.dropdownKeyIndex]._onTabClick();`

File: src/js/controls/Header.js [hunk 3]
```diff
@@ -109,9 +181,9 @@ lm.utils.copy( lm.controls.Header.prototype, {
 	 * @param {lm.item.AbstractContentItem} contentItem
 	 */
 	setActiveContentItem: function( contentItem ) {
-		var i, j, isActive, activeTab;
+		var isActive;
 
-		for( i = 0; i < this.tabs.length; i++ ) {
+		for(var i = 0; i < this.tabs.length; i++ ) {
```
Comment (illumon-public) [inline src/js/controls/Header.js:188]: So you're now declaring this i variable here, but you should actually use `let` to make it a block scoped variable: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let

Using `var` always declares the variable in a global context. So you can do some crazy stuff like this:
```
if (true) {
  var foo = 'this is foo';
  for (var i = 0; i < 10; i++) {
    var foo = 'foo is now ' + i;
  }
}
console.log('foo=', foo, '\ni=', i)
```

Which outputs:
```
foo= foo is now 9 
i= 10
```

File: src/js/controls/Header.js [hunk 4]
```diff
@@ -150,13 +214,212 @@ lm.utils.copy( lm.controls.Header.prototype, {
...
+		this.tabPreviousButton.off('mouseenter', this._handlePreviousMouseEnter);
+		this.tabPreviousButton.off('mouseleave', this._handlePreviousMouseLeave);
+	},
+
+	// on next/previous enter start scroll repeat animation loop
+	// and attach a listener looking for mouseleave
+	// cancel animation on mouse leave, and remove leave listener
+	_handleNextMouseEnter: function() {
+		this.controlsContainer.on('mouseleave', this._handleNextMouseLeave);
+		this._handleScrollRepeat('right', this.tabsContainer.scrollLeft());
```
Comment (illumon-public) [inline src/js/controls/Header.js]: Would make sense to have 'left' and 'right' declared as constants

File: src/js/controls/Header.js [hunk 5]
```diff
@@ -150,13 +214,212 @@ lm.utils.copy( lm.controls.Header.prototype, {
...
+	_handlePreviousMouseLeave: function() {
+		this.rAF = window.cancelAnimationFrame(this.rAF);
+		this.tabPreviousButton.off('mouseleave', this._handlePreviousMouseLeave);
+	},
+
+	// scroll one tab to the right on mouse down
+	// start scrollRepeat if mouse is held down
+	_handleNextButtonMouseDown: function() {
+		var rightOffscreenChild;
+		for (var i = 0; i <  this.tabs.length; i += 1) {
```
Comment (illumon-public) [inline src/js/controls/Header.js]: Double space

File: src/js/controls/Header.js [hunk 6]
```diff
@@ -266,31 +545,150 @@ lm.utils.copy( lm.controls.Header.prototype, {
...
+			this._hideAdditionalTabsDropdown();
+			this.tabs[this.dropdownKeyIndex]._onTabClick();
+			return;
+		}
+		
+		var newIndex = -1;
+		if (e.key === 'ArrowDown') {
+			this.tabDropdownList.eq(this.dropdownKeyIndex).removeClass('lm_keyboard_active');
+			// find the next visible index
+			for (i = this.dropdownKeyIndex + 1 ; i < this.tabDropdownList.length; i++) {
```
Comment (illumon-public) [inline src/js/controls/Header.js]: Could optimize this bit...

```
function getNextDropdownIndex(startIndex, delta) {
  if (this.tabDropdownList.length < 2) {
    return -1;
  }
  let i = (startIndex + delta + this.tabDropdownList.length) % this.tabDropdownList.length;
  while (i != startIndex) {
    if ( this.tabDropdownList.eq(i).css('display') !== 'none' ) {
      return i;
    }
    i = (i + delta + this.tabDropdownList.length) % this.tabDropdownList.length;
  }

  return startIndex;
}
```
Then just:
```
if (e.key === 'ArrowDown') {
  this.dropdownKeyIndex = getNextDropdownIndex(this.dropdownKeyIndex, 1);
} else if (e.key === 'ArrowUp') {
  this.dropdownKeyIndex = getNextDropdownIndex(this.dropdownKeyIndex, -1);
}
```

**Review by illumon-public — COMMENTED** [review]
I'd be alright merging this, though I have a couple optimization suggestions you should probably update the dark theme as well.

File: src/js/controls/Header.js [hunk 7]
```diff
@@ -187,7 +450,14 @@ lm.utils.copy( lm.controls.Header.prototype, {
 		for( var i = 0; i < this.tabs.length; i++ ) {
 			this.tabs[ i ]._$destroy();
 		}
-    	$( document ).off('mouseup', this.hideAdditionalTabsDropdown);
+
+		this._detachWheelListener();
+		this._handleItemPickedDropped();
+
```
Comment (illumon-public) [inline src/js/controls/Header.js:458]: Cancel animation frame here as well

File: src/js/controls/Header.js (hunk 3 above)
Comment (dsmmcken) [reply src/js/controls/Header.js:188]: It's still scoped to the setActiveContent function as it was before, not global. Also we don't have babel here, so I didn't want to go all modern.

File: src/js/controls/Header.js (hunk 7 above)
Comment (dsmmcken) [reply src/js/controls/Header.js:458]: It's already cancelled inside the itemdropped function.

File: src/js/controls/Header.js (hunk 6 above)
Comment (dsmmcken) [reply src/js/controls/Header.js]: that's much more clever.

**Review by illumon-public — APPROVED** [review]

**Pull request merged**